proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+with an error `ApplicationFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The application fees minus rebates and minus overpayment are transferred to
+the respecitve accounts lamport balance, from where the owner can collect them.
+
+### Configuration
+
+Application fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+The ledger will need to store the amount of lamports required for every account.
+Hence changes should be constrained like any account write to the owner of the
+account. All programs will need to implement special instructions so that the
+authority of the accounts can sign to update application fees on the accounts.
+
+Setting the amount to `0` disables application fees. The maximum amount of
+application fees that can be set for an account will be limited to a constant
+amount of SOL (e.g. 20 SOL) so that accounts cannot become inaccessible through
+simple programming errors.
+
+### Rebate
+
+The owner of the account can issue a rebate of the application fees paid to
+write lock a specific account. Similar to the configuration this will need to be
+exposed through special instructions so that programs can implement custom
+authorization and delegate the decision to composing programs.
+
+Simple rebate schemes will verify merely signers, e.g an oracle preventing 3rd
+parties from write-locking their price feed. Dexes will need more complex rebate
+schemes based on instruction sysvar introspection.
+
+Rebate takes the amount of lamports to be rebated, and account on which rebate
+is issued as input. In case of multiple rebates from the same account only the
+highest amount of rebate will be taken into account. The rebated amount is
+always the minimum of rebate issued by the program and the application fees on
+the account. If program rebates `U64::MAX` it means all the application fees on
+the account are rebated. The rebate amount cannot be negative.
+
+### Looking at common cases
+
+#### No application fees enabled
+
+* A payer does not include `PayApplicationFees` in the transaction. The
+  transaction does not write lock any accounts with application fees. Then the
+  transaction is executed without the application fee feature. The payer ends up
+  paying other fees.
+
+* A payer includes `PayApplicationFees(app fees)` in the transaction but none of
+  the accounts have any application fees. This case is considered an overpay
+  case. The payer balance is checked for `other fees + app fees`.
+  1. The payer does not have enough balance: Transaction fails with an error
+    `Insufficient Balance` and the transaction is not even scheduled for
+    execution.
+  2. The payer has enough balance then the transaction is executed and application
+    fees paid are transferred back to the payer in any case.
```

Comment:
Since there is no dApp involved in Application Fee, the refund of paid app fee will be handled by runtime?